### PR TITLE
[compiler][ez] Add more Array.prototype methods

### PR DESCRIPTION
**Pull request by mofeiZ**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30078
* #30077
* #30076
* __->__ #30075
* #30074
* #29891
* #29878

Adds Array.prototype methods that return primitives or other arrays -- naive type inference can be really helpful in reducing mutable ranges -> achieving higher quality memoization.
Also copies Array.prototype methods to our mixed read-only JSON-like object shape.

(Inspired after going through some suboptimal internal compilation outputs.)

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/HIR/ObjectShape.ts [hunk 1]
```diff
@@ -218,6 +218,36 @@ addObject(BUILTIN_SHAPES, BuiltInPropsId, [
...
+      returnValueKind: ValueKind.Primitive,
+    }),
+  ],
+  [
+    "pop",
+    addFunction(BUILTIN_SHAPES, [], {
+      positionalParams: [],
+      restParam: null,
+      returnType: { kind: "Poly" },
+      calleeEffect: Effect.Store,
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/ObjectShape.ts:247]: Similar to `push`, we know that `pop` does not perform interior mutability. `Effect.Store` here models pop similar to `const arr = []; /*...*/; arr[2] = undefined;` (writing exclusively to the container, not to any internal elements)

File: compiler/packages/babel-plugin-react-compiler/src/HIR/ObjectShape.ts [hunk 2]
```diff
@@ -478,6 +521,90 @@ addObject(BUILTIN_SHAPES, BuiltInMixedReadonlyId, [
       noAlias: true,
     }),
   ],
+  [
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/ObjectShape.ts:524]: pasted these to `BuiltInMixedReadonlyId`, which is our read-only JSON-like object kind.

Happy to remove these and refactor later if we prefer to not have duplicates

File: compiler/packages/babel-plugin-react-compiler/src/HIR/ObjectShape.ts (hunk 1 above)
Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/ObjectShape.ts:247]: clever!

File: compiler/packages/babel-plugin-react-compiler/src/HIR/ObjectShape.ts (hunk 2 above)
Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/ObjectShape.ts:524]: this is great!

**Review by josephsavona — APPROVED** [review]
yesssss thank you

**Pull request merged**